python/runtime/db.py

```diff
@@ -25,50 +25,85 @@ def connect_with_data_source(driver_dsn):
     # FIXME(typhoonzero): Should use correct dtype here.
+    null_feature_error = ValueError(
+        "column %s value is NULL, expected dense vector with delimiter %s" %
+        (feature_name, feature_spec["delimiter"]))
     if feature_spec["is_sparse"]:
         if feature_spec["format"] == "kv":
-            items = raw_val.split()
-            items = [item.split(':', 2) for item in items]
-            indices = np.array([int(item[0]) for item in items],
-                               dtype=np.int64)
-            values = np.array([float(item[1]) for item in items],
-                              dtype=np.float32)
+            if is_xgboost and raw_val is None:
+                indices = np.array([], dtype=np.int64)
+                values = np.array([], dtype=np.float32)
+            else:
+                items = raw_val.split()
+                items = [item.split(':', 2) for item in items]
+                indices = np.array([int(item[0]) for item in items],
+                                   dtype=np.int64)
+                values = np.array([float(item[1]) for item in items],
+                                  dtype=np.float32)
         else:
-            indices = np.fromstring(raw_val,
-                                    dtype=int,
-                                    sep=feature_spec["delimiter"])
-            indices = indices.reshape(indices.size, 1)
-            values = np.ones([indices.size], dtype=np.int64)
+            if is_xgboost and raw_val is None:
+                indices = np.array([], dtype=int)
+                values = np.array([], dtype=np.int64)
+            else:
+                indices = np.fromstring(raw_val,
+                                        dtype=int,
+                                        sep=feature_spec["delimiter"])
+                indices = indices.reshape(indices.size, 1)
+                values = np.ones([indices.size], dtype=np.int64)
 
         dense_shape = np.array(feature_spec["shape"], dtype=np.int64)
         return indices, values, dense_shape
     elif feature_spec["delimiter"] != "":
         # Dense string vector
         if feature_spec["dtype"] == "float32":
-            vec = np.fromstring(raw_val,
-                                dtype=np.float32,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                raise null_feature_error
+            else:
+                vec = np.fromstring(raw_val,
+                                    dtype=np.float32,
+                                    sep=feature_spec["delimiter"])
         elif feature_spec["dtype"] == "int64":
-            vec = np.fromstring(raw_val,
-                                dtype=np.int64,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                raise null_feature_error
+            else:
+                vec = np.fromstring(raw_val,
```

Comment:
How can we ensure that each item of a specific dense feature has the same shape?
np.fromstring may returns various shape for different input strings.

Reply:
This is done by feature derivation before training.